### PR TITLE
SQL: Concat should be always not nullable

### LINKED ISSUE #36169: SQL: CONCAT should return NULL when the field involved has a NULL value?
SQL Server 2017, by default, will consider [`null` values in `CONCAT()` function as empty strings](https://docs.microsoft.com/en-us/sql/t-sql/functions/concat-transact-sql?view=sql-server-2017) and return a non-null value. But, it also offers the option of changing this setting by using `CONCAT_NULL_YIELDS_NULL` setting. They do warn, though, that this option will be `ON` in a future version of SQL Server: https://docs.microsoft.com/en-us/sql/t-sql/statements/set-concat-null-yields-null-transact-sql?view=sql-server-ver15. The [preview docs](https://docs.microsoft.com/en-us/sql/t-sql/statements/set-concat-null-yields-null-transact-sql?view=sql-server-ver15) for SQL Server 2019 does still have the option default as `OFF` (so `null` is considered empty string).

PostgreSQL will also "ignore" null values and treat them as empty strings: https://www.postgresql.org/docs/11/functions-string.html

### PR DESCRIPTION
**Pull request by astefan**
This PR fixes the CONCAT function case where, if one of the values is `null`, the function returns `null`. Instead it should always consider `null` as an empty string. Fixes https://github.com/elastic/elasticsearch/issues/36169.

### CODE REVIEW

**Review by matriv — COMMENTED** [review]
I'd prefer to have an optimizer test for that, actually QueryFolder test to check that the concat is not eliminated because of the FoldNull() rule.

We can keep the integ test too, of course.

**Review by costin — COMMENTED** [review]
+1 to what Marios said. The advantage of that unit tests is that it checks the behavior in isolation and thus we can identify in the future if there are any side effects.

**Comment by astefan** [comment]
run the gradle build tests 1

**Comment by matriv** [comment]
@astefan The test you added is also nice, but would be great to also add one similar to those: https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/optimizer/OptimizerTests.java#L407 that checks that FoldNull() doesn't fold the concat() function.

Sorry for not being more specific before and I confused you with mentioning the `QueryFolderTests`.

Thank you!

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/optimizer/OptimizerTests.java [hunk 1]
```diff
@@ -111,6 +112,8 @@
     private static final Literal FOUR = L(4);
     private static final Literal FIVE = L(5);
     private static final Literal SIX = L(6);
+    
+    private static final String EMPTY_STRING = "";
```
Comment (matriv) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/optimizer/OptimizerTests.java]: Is it already available in TestUtils?

**Review by matriv — APPROVED** [review]
LGTM. Thx for the tests!

**Pull request merged**